### DISCUSSION TITLE
Are there any plans for mongoose to switch to ECMAScript Modules (ESM)?

### DISCUSSION THREAD

**Discussion by hasezoey** [Automattic/mongoose#11929] [Q&A]
As the title says, i wanted to ask if there are any plans for mongoose to switch to ESM in the next major versions, because NodeJS 12 supports ESM without flags and mongoose uses this as the minimal NodeJS version and ESM seems to be preferred now over CJS.

Note: this is just a question if there are any plans / consideration for ESM, this is not a feature proposal.

**Comment by Uzlopak** [comment]
It should be possible to import mongoose as esm module. Is this not "enough"?

What would be needed to fulfill this feature request?

**Reply by hasezoey** [reply]
> What would be needed to fulfill this feature request?

like i said in the Original Post "this is not a feature proposal"

---

sorry if i was not clear enough, i meant to ask about the mongoose source code, if there are any plans to switch to ESM (eg from `require & module.exports` to `import & export`), this is also a question just out of curiosity, not because i need it

**Reply by Uzlopak** [reply]
Nah, dont worry about my wording. It is easier to handle this as a feature request even if it is not a feature request. If I know what changes we have to do, it makes it possible to estimate or to decide what the best approach would be.

My idea regarding this is, that we should make use of webpack, rollup or maybe even better use vite and bundle the javascript code into a single files. And of those exports we could export it as esm. It should also improve the loading time when only one files needs to be hoisted. And we could make use of the module-entry in the package.json.

**Reply by vkarpov15** [reply]
No plans currently. We're not opposed to switching to ESM eventually, but right now we have more pressing priorities - performance improvements, bug fixes, TypeScript improvements, etc.
